RFC-0008-plan-checker.md

```diff
@@ -0,0 +1,137 @@
+what is supported between both evaluation engines.  Because of these underlying differences, the optimizer may generate plans that
+can't be executed in C++ clusters, or likewise, a Presto Java cluster could be misconfigured to generate plans that only work with
+C++ clusters.
+
+If a plan is generated that can't be executed in the cluster, the query will fail with an error message during the query execution
+phase.  This is a poor user experience, as the user has to wait for the query to fail before they can take corrective action, and
+this might have occurred after lengthy queueing.  Additionally, this failure would occur after worker resources have already been
+allocated to the query, which is wasteful.  This RFC proposes a plan checker that can be run before the query is executed to ensure
+a quick validation of the plan.
+
+### [Optional] Goals
+
+* Provide a mechanism to validate Presto to Velox plan conversion during planning phase
+* Add an SPI to allow custom validators to be added to suit individual business needs
+
+### [Optional] Non-goals
+
+* Ensure all Velox plans are executable in a Velox cluster--many checks are done at runtime and may not be caught by the plan checker
+
+## Proposed Implementation
+
+The Presto runtime centralizes plan validation logic into the `PlanChecker` class.  There exist three phases to this class: 
+
+* `validateIntermediatePlan`
+* `validateFinalPlan`
+* `validateFragmentedPlan`
+
+It is generally useful to allow this class to be configured with an SPI that allows for custom plan validation logic to be added.
+For example, a business may decide that a certain type is not allowed, or add a check to ensure that plans that are overly
+complicated are killed.
+
+An SPI will be added that will add more checks to the `PlanChecker` class.  The SPI will contain a field indicating which phase
+of the plan checker to be added, and a validator that will be run during that phase.  The SPI will be added to the `PlanChecker`
+class, and the `PlanChecker` class will be modified to run the SPIs during the appropriate phase.
+
+A new endpoint to the Presto sidecar will be added that will attempt to convert a Presto plan to a Velox plan.  In the
+`presto-native-plugin` module, a new implementation of the SPI will be added which will add a check to the `PlanChecker` class
+which will call the sidecar to attempt to convert the plan.  If the conversion fails, the plan checker will fail the plan.  It will
+be added at the `validateFragmentedPlan` phase--this is because it's not until plan fragmentation occurs that we know which portions
+of the plan will be executed in the coordinator, and which will be executed in the workers.
+
+The code for the plan checker will run `PrestoToVeloxQueryPlan`, which is used by the workers to convert the Presto plan fragment
+to a Velox plan fragment.  If the conversion fails, the plan checker will fail the plan, returning with it the reason for the failure.
```

Comment:
It might be obvious, but just to add that since the plan checker is a plugin, it is not run at all if the plugin is not present

Reply:
I've added this to the adoption plan.